src/smclient/smclient.hpp

```diff
@@ -111,10 +131,43 @@ class SMClient : public sm::launcher::InstanceStatusReceiverItf,
     void OnClockUnsynced() override;
 
 private:
-    static constexpr auto cOpenPort   = CONFIG_AOS_SM_OPEN_PORT;
-    static constexpr auto cSecurePort = CONFIG_AOS_SM_SECURE_PORT;
-
-    OpenHandler mOpenHandler;
+    static constexpr auto cOpenPort                 = CONFIG_AOS_SM_OPEN_PORT;
+    static constexpr auto cSecurePort               = CONFIG_AOS_SM_SECURE_PORT;
+    static constexpr auto cMaxConnectionSubscribers = 2;
+
+    void  OnConnect() override;
+    void  OnDisconnect() override;
+    Error ReceiveMessage(const Array<uint8_t>& data) override;
+
+    void  UpdatePBHandlerState();
+    Error SendNodeConfigStatus(const String& version, const Error& configErr);
+    Error ProcessGetNodeConfigStatus(const servicemanager_v4_GetNodeConfigStatus& pbGetNodeConfigStatus);
+    Error ProcessCheckNodeConfig(const servicemanager_v4_CheckNodeConfig& pbCheckNodeConfig);
+    Error ProcessSetNodeConfig(const servicemanager_v4_SetNodeConfig& pbSetNodeConfig);
+    Error ProcessGetAverageMonitoring(const servicemanager_v4_GetAverageMonitoring& pbGetAverageMonitoring);
+    Error ProcessRunInstances(const servicemanager_v4_RunInstances& pbRunInstances);
+    Error ProcessImageContentInfo(const servicemanager_v4_ImageContentInfo& pbImageContentInfo);
+    Error ProcessImageContent(const servicemanager_v4_ImageContent& pbImageContent);
+
+    OpenHandler                                 mOpenHandler;
+    iam::nodeinfoprovider::NodeInfoProviderItf* mNodeInfoProvider {};
+    sm::launcher::LauncherItf*                  mLauncher {};
+    sm::resourcemanager::ResourceManagerItf*    mResourceManager {};
+    aos::monitoring::ResourceMonitorItf*        mResourceMonitor {};
+    downloader::DownloadReceiverItf*            mDownloader {};
+    communication::ChannelManagerItf*           mChannelManager {};
+
+    Mutex mMutex;
+    bool  mClockSynced = false;
+    bool  mProvisioned = false;
+
+    StaticArray<ConnectionSubscriberItf*, cMaxConnectionSubscribers> mConnectionSubscribers;
```

Comment:
Why don't you use aos_core_lib map?
Anyway, current implementation doesn't check if subscribe has been already added to the StaticArray.

Reply:
If `ConnectionSubscriberItf*` is the key for map, what will be map value in this case?